include/nbl/system/IFileArchive.h

```diff
@@ -6,32 +6,213 @@
+		system::path name;
+
+		//! The name of the file including the path
+		/** If this is a file or folder in the virtual filesystem and the archive was
+		created with the ignoreDirs flag then it will be the same as Name. */
+		system::path fullName;
+
+		//! The size of the file in bytes
+		uint32_t size;
+
+		//! The ID of the file in an archive
+		/** This is used to link the FileList entry to extra info held about this
+		file in an archive, which can hold things like data offset and CRC. */
+		uint32_t ID;
+
+		//! FileOffset inside an archive
+		uint32_t offset;
+
+		//! True if this is a folder, false if not.
+		bool isDirectory;
+
+		
+		E_ALLOCATOR_TYPE allocatorType;
+
+		//! The == operator is provided so that CFileList can slowly search the list!
+		inline bool operator ==(const struct SFileListEntry& other) const
+		{
+			if (isDirectory != other.isDirectory)
+				return false;
+
+			return core::strcmpi(fullName.string(), other.fullName.string()) == 0;
+		}
+
+		//! The < operator is provided so that CFileList can sort and quickly search the list.
+		inline bool operator <(const struct SFileListEntry& other) const
+		{
+			if (isDirectory != other.isDirectory)
+				return isDirectory;
+
+			return fullName < other.fullName;
+		}
+	};
 	struct SOpenFileParams
 	{
 		std::filesystem::path filename;
 		std::string_view password;
 	};
 
-	//! Opens a file based on its name
-	virtual core::smart_refctd_ptr<IFile> readFile(const SOpenFileParams& params) = 0;
+	std::mutex fileMutex;
+	std::atomic_flag* m_fileFlags;
+	std::byte* m_filesBuffer = nullptr;
+	
+	core::smart_refctd_ptr<IFile> readFile(const SOpenFileParams& params)
+	{
+		auto index = getIndexByPath(params.filename);
+		switch (this->getFileType(index))
+		{
+		case EAT_NULL:
+			return getFile_impl<CNullAllocator>(params, index);
+			break;
+		case EAT_MALLOC:
+			return getFile_impl<CPlainHeapAllocator>(params, index);
+			break;
+		case EAT_VIRTUAL_ALLOC:
+			return getFile_impl<CFileViewVirtualAllocatorWin32>(params, index); //TODO linux
+			break;
+		}
+		assert(false);
+		return nullptr;
+	}
+	virtual core::smart_refctd_ptr<IFile> readFile_impl(const SOpenFileParams& params) = 0;
+	int32_t getIndexByPath(const system::path& p)
+	{
+		for (int i = 0; i < m_files.size(); ++i)
+		{
+			if (p == m_files[i].fullName) return i;
+		}
+		return -1;
+	}
+	E_ALLOCATOR_TYPE getFileType(uint32_t index)
+	{
+		return m_files[index].allocatorType;
+	}
+	template<class Allocator>
+	core::smart_refctd_ptr<CInnerArchiveFile<Allocator>> getFile_impl(const SOpenFileParams& params, const uint32_t index)
+	{
+		static const size_t MAX_SIZE = std::max(sizeof(CInnerArchiveFile<CPlainHeapAllocator>), sizeof(CInnerArchiveFile<CFileViewVirtualAllocatorWin32>));
+		std::unique_lock lock(fileMutex);
+
+		auto* file = reinterpret_cast<CInnerArchiveFile<Allocator>*>(m_filesBuffer + index * MAX_SIZE);
+		const auto oldRefcount = file->grab();
```

Comment:
please add comment that you intentionally call grab() on maybe-not-existing object